service/history/replicationTaskFetcher.go

```diff
@@ -289,18 +287,24 @@ func (f *ReplicationTaskFetcherImpl) getMessages(
 	response, err := f.remotePeer.GetReplicationMessages(ctx, request)
 	if err != nil {
 		f.logger.Error("Failed to get replication tasks", tag.Error(err))
+		for _, req := range f.requestByShard {
+			close(req.respChan)
+		}
+		f.requestByShard = make(map[int32]*replicationTaskRequest, requestsMapSize)
```

Comment:
[Nit] should we just put this in defer statement near the top of the function?

Reply:
> Is it possible to put this in defer statement near the top of the function?

yes, there can be a catch all defer which close all channels.
i am not doing the suggestion here because the `len(f.requestByShard)` can be large, meaning that on the happy path, the iteration will be done twice.